tests/websocket.c

```diff
@@ -28,147 +48,412 @@ struct worker_thread {
 	int msg_received;
 	int msg_sent;
 	int byte_count;
+	int id;
 	pthread_t thread;
-
-	struct evbuffer *buffer;
+	enum worker_state state;
+	int timeout_seconds;
+
+	/* non-encoded websocket key */
+	char ws_key[16];
+	/* expected response */
+	char ws_response[28];
+	size_t ws_response_len;
+	/* actual response */
+	char *sec_websocket_accept;
+	/* masking */
+	enum mask_config mask_cfg;
+	int mask_applied;
+
+	/* current header */
+	char *cur_hdr_key;
+	size_t cur_hdr_key_len; /* not including trailing \0 */
+	char *cur_hdr_val;
+	size_t cur_hdr_val_len; /* not including trailing \0 */
+	int hdr_last_cb_was_name; /* tells us if the last call was header name or value */
+
+	struct evbuffer *rbuffer;
 	int got_header;
 
+	struct evbuffer *wbuffer;
+
 	int verbose;
+	int fd;
+	struct event ev_r;
 	struct event ev_w;
+
+	http_parser parser;
+	http_parser_settings settings;
+
+	int (*debug)(const char *fmt, ...);
 };
 
+int debug_noop(const char *fmt, ...) {
+	(void)fmt;
+	return 0;
+}
+
+int debug_verbose(const char *fmt, ...) {
+	int ret;
+	va_list vargs;
+	va_start(vargs, fmt);
+	ret = vfprintf(stderr, fmt, vargs);
+	va_end(vargs);
+	return ret;
+}
+
 void
-process_message(struct worker_thread *wt, size_t sz) {
+hex_dump(struct worker_thread *wt, char *p, size_t sz) {
+	wt->debug("hex dump of %p (%ld bytes)\n", p, sz);
+	for (char *cur = p; cur < p + sz; cur += 16) {
+		char letters[16] = {0};
+		int limit = (cur + 16) > p + sz ? (sz % 16) : 16;
+		wt->debug("%08lx ", cur - p); /* address */
+		for (int i = 0; i < limit; i++) {
+			wt->debug("%02x ", (unsigned int)(cur[i] & 0xff));
+			letters[i] = isprint(cur[i]) ? cur[i] : '.';
+		}
+		for (int i = limit; i < 16; i++) { /* pad on last line */
+			wt->debug("   "); /* 3 spaces for "%02x " */
+		}
+		wt->debug(" %.*s\n", limit, letters);
+	}
+}
+
+void
+evbuffer_debug_dump(struct worker_thread *wt, struct evbuffer *buffer) {
+	size_t sz = evbuffer_get_length(buffer);
+	char *data = malloc(sz);
+	if (!data) {
+		fprintf(stderr, "failed to allocate %ld bytes\n", sz);
+		return;
+	}
+	evbuffer_remove(buffer, data, sz);
+	hex_dump(wt, data, sz);
+	evbuffer_prepend(buffer, data, sz);
+	free(data);
+}
+
+static void
+wait_for_possible_read(struct worker_thread *wt);
+
+static void
+wait_for_possible_write(struct worker_thread *wt);
+
+static void
+ws_enqueue_frame(struct worker_thread *wt);
 
-	// printf("process_message\n");
-	if(wt->msg_received % 10000 == 0) {
-		printf("thread %u: %8d messages left (got %9d bytes so far).\n",
-			(unsigned int)wt->thread,
-			wt->msg_target - wt->msg_received, wt->byte_count);
+void
+process_message(struct worker_thread *wt, size_t sz) {
+	if (wt->msg_received && wt->msg_received % 1000 == 0) {
+		printf("thread %d: %8d messages left (got %9d bytes so far).\n",
+		       wt->id,
+		       wt->msg_target - wt->msg_received, wt->byte_count);
```

Comment:
sorry I didn't think about it earlier, but this might make more sense as a sleeping thread that reports at a fixed frequency.

Reply:
That's added! It took me a while due to the deprecation of POSIX semaphores on macOS but it works on both Mac and Linux now.